### PR TITLE
User-facing WattTime client errors

### PR DESCRIPTION
**Pull request by bderusha**
Issue Number: 204

## Summary
Surfaces WattTime errors to the API user.

## Changes

The WebAPI uses the [.Net controller filter pipeline](https://docs.microsoft.com/en-us/aspnet/core/mvc/controllers/filters?view=aspnetcore-6.0) to ensure that all requests respond with a consistent JSON schema.

.Net controller filter pipeline image

Controllers are responsible for managing the "Success" responses.  If an error occurs an exception will be thrown and the [custom Exception Filter](./Filters/HttpResponseExceptionFilter.cs) will manage converting that exception into the appropriate JSON response.

The error schema used by this API is that of the [ValidationProblemDetails](https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.mvc.validationproblemdetails?view=aspnetcore-6.0) class in order to provide a consistent error schema while taking advantage of errors automatically returned by the framework.

## Checklist

- [x] Local Tests Passing?
- [x] CICD and Pipeline Tests Passing?
- [x] Added any new Tests?
- [x] Documentation Updates Made?
- [ ] This is not a breaking change. If it is, please describe it below.

## Is this a breaking change?
If yes, what workflow does this break? 

Changes the API.yml error schema

### CODE REVIEW

File: src/dotnet/CarbonAware.WebApi/README.md [hunk 1]
```diff
@@ -77,4 +77,15 @@ EG
   "embodiedEmissionsValue": null,
   "functionalUnitValue": null
 }
-```
\ No newline at end of file
+```
+
+
+## Error Handling
```
Comment (JenMadiedo) [inline src/dotnet/CarbonAware.WebApi/README.md:83]: I think it'd be useful to add an extra sentence here basically along the lines of "if you throw an exception somewhere in the stack but catch it before it hits the controller, then this won't be triggered. If the exception hits the controller than this standardized process takes over."

File: src/dotnet/CarbonAware.WebApi/README.md [hunk 2]
```diff
@@ -77,4 +77,15 @@ EG
...
+
+The WebAPI uses the [.Net controller filter pipeline](https://docs.microsoft.com/en-us/aspnet/core/mvc/controllers/filters?view=aspnetcore-6.0) to ensure that all requests respond with a consistent JSON schema.
```
Comment (JenMadiedo) [inline src/dotnet/CarbonAware.WebApi/README.md:85]: I'm not sure it's if worth expanding slightly here that not only is it consistent within the repo or with dotnet generally.

**Review by JenMadiedo — APPROVED** [review]
Left a handful of comments on the readme but otherwise lgtm (thanks for walking through offline!)

File: src/dotnet/CarbonAware.WebApi/README.md (hunk 2 above)
Comment (KanishkT123) [reply src/dotnet/CarbonAware.WebApi/README.md:85]: I would definitely want more expansion on what "consistent" means here.

File: src/dotnet/CarbonAware.WebApi/Filters/HttpResponseExceptionFilter.cs [hunk 3]
```diff
@@ -0,0 +1,48 @@
+using CarbonAware.Interfaces;
+using Microsoft.AspNetCore.Mvc;
+using Microsoft.AspNetCore.Mvc.Filters;
+using System.Net;
+using System.Diagnostics;
+
+namespace CarbonAware.WebApi.Filters;
+
+public class HttpResponseExceptionFilter : IExceptionFilter
```
Comment (rakker91) [inline src/dotnet/CarbonAware.WebApi/Filters/HttpResponseExceptionFilter.cs:9]: This class needs logging and telemetry

**Review by rakker91 — APPROVED** [review]

File: src/dotnet/CarbonAware.Tools.WattTimeClient/WattTimeClient.cs [hunk 4]
```diff
@@ -262,7 +262,7 @@ private async Task UpdateAuthTokenAsync()
             if (data == null)
             {
                 Log.LogError("Login failed for user {username}.  Response: {response}", this.Configuration.Username, response);
-                throw new WattTimeClientException("Login failed.", response);
+                throw new WattTimeClientHttpException($"Login failed for user: '{this.Configuration.Username}'", response);
```
Comment (gfmatthews) [inline src/dotnet/CarbonAware.Tools.WattTimeClient/WattTimeClient.cs:265]: 🔥 nice!

File: src/dotnet/CarbonAware.WebApi/api.yaml [hunk 5]
```diff
@@ -127,8 +127,16 @@ components:
     CarbonAwareError:
       type: object
       properties:
-        message:
+        type:
```
Comment (gfmatthews) [inline src/dotnet/CarbonAware.WebApi/api.yaml:130]: I think the fact that we have to update this _artisanal hand-crafted API spec_ is perhaps an indicator that we should really prioritize that auto-generation of the spec story that's sitting out there on the backlog

**Review by gfmatthews — APPROVED** [review]
Approved with non-blocking notes

**Pull request merged**